### PR TITLE
Allow configuration of loggers through Caddyfile global options

### PR DESCRIPTION
**Pull request by kujenga**
This PR is a follow up on the previous attempt in https://github.com/caddyserver/caddy/pull/3968 but instead implemented as a global logging option.

It follows up on discussion in https://github.com/caddyserver/caddy/issues/3958 about adding more functionality to the Caddyfile. I also added in another commit for a small new "replace" filter aimed at the same goal as I wrote up in the original ticket, providing more tools to remove sensitive data from logs.

Closes https://github.com/caddyserver/caddy/issues/3958

### CODE REVIEW

File: caddyconfig/httpcaddyfile/builtins.go
Comment (francislavoie) [inline caddyconfig/httpcaddyfile/builtins.go:740]: I don't think we want to add `include` and `exclude` to the `log` directive, but we would want it for the global option. So that would probably need to be wrapped with that boolean.

File: caddyconfig/httpcaddyfile/options_test.go
Comment (francislavoie) [inline caddyconfig/httpcaddyfile/options_test.go:64]: I'm confused about this one - shouldn't this be an error? I don't think it's valid to have a `log` without a name in globals.

File: modules/logging/filters.go
Comment (francislavoie) [inline modules/logging/filters.go:28]: Let's hold off on this for now, we can do that in a separate PR. Keep this PR focused on just the global option, please!

File: caddyconfig/httpcaddyfile/builtins.go
Comment (kujenga) [reply caddyconfig/httpcaddyfile/builtins.go:740]: Makes sense, that was one of the things I was wondering about, I can disable these for the global case.

File: modules/logging/filters.go
Comment (kujenga) [reply modules/logging/filters.go:28]: Sure, happy to do that 👍

File: caddyconfig/httpcaddyfile/options_test.go
Comment (kujenga) [reply caddyconfig/httpcaddyfile/options_test.go:64]: I was going off your comment here about how the name could be optional: https://github.com/caddyserver/caddy/issues/3958#issuecomment-756776300

This is a bit of a strange case because the second block overwrites the first one.

I like the idea of making it name required field, explicit in all cases for the global option. I'll make that change.

File: caddyconfig/httpcaddyfile/builtins.go
Comment (francislavoie) [reply caddyconfig/httpcaddyfile/builtins.go]: To be clear - we want to _allow_ `include` and `exclude` for globals, but _disallow_ it for the `log` directive. The `log` directive's purpose is to configure access logs, so the `include` part of it is already "managed".

File: caddyconfig/httpcaddyfile/options_test.go
Comment (francislavoie) [reply caddyconfig/httpcaddyfile/options_test.go]: Oh I see, well in the case yeah I think it's okay that omitting the name uses `default`, but it should be an error if the same name is used (or implied) more than once. Overriding is non-apparent behaviour and is almost always going to be misconfiguration.

File: caddyconfig/httpcaddyfile/directives.go
Comment (francislavoie) [inline caddyconfig/httpcaddyfile/directives.go:513]: I understand the purpose of this change, but this is technically a breaking change. This is meant to be available to third party plugins so they can register global options of their own.

That said, I don't think any actually do yet because support for it was incomplete (see https://github.com/caddyserver/caddy/issues/3634) but that was resolved in https://github.com/caddyserver/caddy/pull/3990 which will be released in v2.4.0.

So if we need to make this change, we'll have to make the change before the v2.4.0 release to avoid future breakages.

I'm thinking this refactor of `UnmarshalGlobalFunc` might need to also be moved into a separate PR and merged ahead of this one, but I'll let @mholt chime in first.

Comment (kujenga) [reply caddyconfig/httpcaddyfile/directives.go:513]: Ah! Great point.

Without this or something like it, it might be a bit hacky to do the check for duplicate/colliding log configurations, and it would be something that had to be in Caddy itself, i.e. not possible as part of the public API. FWIW, this seems like it may be a beneficial construct to have going forward to make the API more easily extensible in a non-breaking way, like directives work today.

File: caddyconfig/httpcaddyfile/options_test.go
Comment (kujenga) [reply caddyconfig/httpcaddyfile/options_test.go]: Got it, makes sense to me, I'll add back the allowance for making it optional.

File: caddyconfig/httpcaddyfile/directives.go
Comment (kujenga) [reply caddyconfig/httpcaddyfile/directives.go:513]: Hey @mholt @francislavoie , just wanted to check in here, any further thoughts on whether or not this is an acceptable change to make?

If there is an issue with this interface shift let me know and I take another look here.

Comment (mholt) [reply caddyconfig/httpcaddyfile/directives.go:513]: It hasn't been released yet, so we can technically change it, though that's not really desirable.

I don't _love_ the `GlobalHelper` state (we do already have something similar in `Helper`: https://pkg.go.dev/github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile#Helper)

If the only goal is to pass in the state, then maybe that could/should just be another argument to `UnmarsahlGlobalFunc` functions.

Comment (kujenga) [reply caddyconfig/httpcaddyfile/directives.go]: Looking at this again, I'm actually able to do what's needed without the state involved by using a healthy dose of type assertions to look at the "existing values" passed into the UnmarshalGlobalFunc. I've pushed commits for that now that I can squash down if the result looks amenable.

Comment (kujenga) [reply caddyconfig/httpcaddyfile/directives.go]: I went ahead and rebased on master and squashed things down

**Comment by kujenga** [comment]
Hi @mholt @francislavoie , just checking in, any thoughts on the simplified approach this PR now takes? Happy to make any further iterations as wanted.

**Review by francislavoie — APPROVED** [review]
Looks good to me! Thanks for the effort! :tada:

File: caddyconfig/httpcaddyfile/options_test.go
Comment (francislavoie) [inline caddyconfig/httpcaddyfile/options_test.go:10]: This is testing the global options in isolation, which is good, but I think we should also test global logs in combination with logs configured in sites.

An easy way to do that is by writing an adapt test, you can find the ones we have already here: https://github.com/caddyserver/caddy/tree/master/caddytest/integration/caddyfile_adapt

You could maybe move all of the tests in this file to be adapt test files instead, they're a bit easier to read and update, and render a diff of the JSON on error.

**Review by francislavoie — CHANGES_REQUESTED** [review]
Actually, noticed a couple more things on a quick second pass, sorry :sweat_smile:

File: caddyconfig/httpcaddyfile/options_test.go
Comment (kujenga) [reply caddyconfig/httpcaddyfile/options_test.go:10]: Great, thanks for pointing that out. I ended up splitting the tests so that the more complex successful test cases are in the adapt folder, and the failure cases remain unit tests, since it seems like the adapt folder does not support that? Let me know if this matches what you'd expect.

Comment (francislavoie) [reply caddyconfig/httpcaddyfile/options_test.go:10]: Yeah that's perfect, thanks

**Comment by kujenga** [comment]
Alright, this is rebased on master now as well, if there's anything else just let me know!

File: caddyconfig/httpcaddyfile/options_test.go
Comment (francislavoie) [reply caddyconfig/httpcaddyfile/options_test.go:10]: I think we're still missing an adapt test that has both a global logger and a logger in a site. Something like this:

```
{
	log {
		output file caddy.log
	}
}

:8884 {
	log {
		output file access.log
	}
}
```

Comment (kujenga) [reply caddyconfig/httpcaddyfile/options_test.go:10]: Great idea, and it turns out that caught a subtle edge-case bug the default logger would end up with logs configured for it excluded by this logic: https://github.com/caddyserver/caddy/blob/51f35ba03f78e691d0e141b24ff19f82bfb34b67/caddyconfig/httpcaddyfile/httptype.go#L313-L322

That case is now handled better and a new test case file covers it.

Comment (francislavoie) [reply caddyconfig/httpcaddyfile/options_test.go:10]: Perfect, thank you!

**Review by francislavoie — APPROVED** [review]

File: caddyconfig/httpcaddyfile/builtins.go
Comment (mholt) [inline caddyconfig/httpcaddyfile/builtins.go:629]: Will `globalOption` ever be `true` while `globalLogNames` is nil? If not, we can drop the `globalOption` param and just assume that if `globalLogNames` is non-nil, `globalOption` must be true.

**Review by mholt — APPROVED** [review]
Awesome, thanks for your reviews Francis and for the work on this, @kujenga -- LGTM, just one comment where a slight improvement could be made if I'm right. Ignore if not. :)

File: caddyconfig/httpcaddyfile/builtins.go
Comment (kujenga) [reply caddyconfig/httpcaddyfile/builtins.go]: Good point! I think that was an artifact of an earlier iteration, I collapsed the parameters and added an internal variable to keep the logic explicit.

**Comment by kujenga** [comment]
Alright, this has been rebased on master and I believe the last comment is now addressed!

**Review by francislavoie — APPROVED** [review]

**Review by mholt — APPROVED** [review]
Excellent, thanks for the contribution @kujenga ! Please feel free to contribute again sometime. :) Especially if any issues with these changes are reported later. Thank you!

**Comment by kujenga** [comment]
Awesome, thank you! Feel free to ping me if anything does come up 👍

**Comment by francislavoie** [comment]
@kujenga would you like to take a shot at writing the docs for this? You can find them here https://github.com/caddyserver/website/blob/master/src/docs/markdown/caddyfile/options.md

**Pull request merged**
